src/fparser/Fortran2003.py

```diff
@@ -4991,184 +5002,327 @@ class End_Select_Type_Stmt(EndStmtBase): # R824
+class Nonlabel_Do_Stmt(StmtBase, WORDClsBase):  # pylint: disable=invalid-name
     """
+    R829
+
     <nonlabel-do-stmt> = [ <do-construct-name> : ] DO [ <loop-control> ]
     """
     subclass_names = []
     use_names = ['Do_Construct_Name', 'Loop_Control']
-    def match(string): return WORDClsBase.match('DO', Loop_Control, string)
-    match = staticmethod(match)
 
-class Loop_Control(Base): # R830
+    @staticmethod
+    def match(string):
+        '''
+        :param str string: Fortran code to check for a match
+        :return: code line matching the nonlabeled "DO" statement
+        :rtype: string
+        '''
+        return WORDClsBase.match('DO', Loop_Control, string)
+
+
+class Loop_Control(Base):  # pylint: disable=invalid-name
     """
-    <loop-control> = [ , ] <do-variable> = <scalar-int-expr> , <scalar-int-expr> [ , <scalar-int-expr> ]
+    R830
+
+    <loop-control> = [ , ] <do-variable> = scalar-int-expr,
+                                           scalar-int-expr
+                                           [ , <scalar-int-expr> ]
                      | [ , ] WHILE ( <scalar-logical-expr> )
     """
     subclass_names = []
     use_names = ['Do_Variable', 'Scalar_Int_Expr', 'Scalar_Logical_Expr']
+
+    @staticmethod
     def match(string):
+        '''
+        :param str string: Fortran code to check for a match
+        :return: 3-tuple containing strings and instances of the classes
+                 determining loop control (optional comma delimiter,
+                 optional scalar logical expression describing "WHILE"
+                 condition or optional counter expression containing loop
+                 counter and scalar integer expression)
+        :rtype: 3-tuple of objects or nothing for an "infinite loop"
+        '''
+        # pylint: disable=unbalanced-tuple-unpacking
+        optional_delim = None
+        # Match optional delimiter
         if string.startswith(','):
             line, repmap = string_replace_map(string[1:].lstrip())
+            optional_delim = ", "
         else:
             line, repmap = string_replace_map(string)
-        if line[:5].upper()=='WHILE' and line[5:].lstrip().startswith('('):
-            l = line[5:].lstrip()
-            i = l.find(')')
-            if i!=-1 and i==len(l)-1:
-                return Scalar_Logical_Expr(repmap(l[1:i].strip())),
-        if line.count('=')!=1: return
-        var,rhs = line.split('=')
+        # Match "WHILE" scalar logical expression
+        if line[:5].upper() == 'WHILE' and line[5:].lstrip().startswith('('):
+            lbrak = line[5:].lstrip()
+            i = lbrak.find(')')
+            if i != -1 and i == len(lbrak)-1:
+                scalar_logical_expr = \
+                    Scalar_Logical_Expr(repmap(lbrak[1:i].strip()))
+                return scalar_logical_expr, None, optional_delim
+        # Match counter expression
+        if line.count('=') != 1:
+            return
```

Comment:
Please could you add (to) a test in order to cover this line.

Reply:
Added.